### PR TITLE
PPS Full simulation + PPS proton transport with totem parameterization

### PR DESCRIPTION
**Pull request by mundim**
PR related to the presentation in the Simulation meeting of June, 8th. It includes the first PR with full simulation of PPS, including the new SD for the tracker and timing detectors. It relies on the new proton transport, redesigned in order to allow the propagation using also the TOTEM parameterization for the realistic optics and HECTOR as well. It also includes a utility module that holds methods used by simulation and reconstruction (full and fast as well), isolated into a separated module to avoid duplication of maintenance efforts.

@fabferro @jan-kaspar @dilsonjd @nminafra @forthommel @grzanka @setesami @nogima

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@cmsbuild please test

File: Configuration/StandardSequences/python/DigiToRaw_cff.py [hunk 1]
```diff
@@ -17,11 +17,12 @@
 from EventFilter.CastorRawToDigi.CastorDigiToRaw_cfi import *
 from EventFilter.RawDataCollector.rawDataCollector_cfi import *
 from L1Trigger.Configuration.L1TDigiToRaw_cff import *
+from EventFilter.CTPPSRawToDigi.ctppsDigiToRaw_cfi import *
 
 from Configuration.ProcessModifiers.premix_stage1_cff import premix_stage1
 
 #DigiToRaw = cms.Sequence(L1TDigiToRaw*siPixelRawData*SiStripDigiToRaw*ecalPacker*esDigiToRaw*hcalRawData*cscpacker*dtpacker*rpcpacker*rawDataCollector)
-DigiToRaw = cms.Sequence(L1TDigiToRaw*siPixelRawData*SiStripDigiToRaw*ecalPacker*esDigiToRaw*hcalRawData*cscpacker*dtpacker*rpcpacker*castorRawData*rawDataCollector)
+DigiToRaw = cms.Sequence(L1TDigiToRaw*siPixelRawData*SiStripDigiToRaw*ecalPacker*esDigiToRaw*hcalRawData*cscpacker*dtpacker*rpcpacker*ctppsRawData*castorRawData*rawDataCollector)
```
Comment (slava77) [inline Configuration/StandardSequences/python/DigiToRaw_cff.py]: add this with ```ctpps_2016```
see e.g. https://github.com/cms-sw/cmssw/blob/b59dc5246fb83a92d5f9cf445a6f2891d6181a32/Configuration/StandardSequences/python/RawToDigi_cff.py#L98-L102

File: Configuration/StandardSequences/python/ctppsRecoForSimulation2016_cff.py [hunk 2]
```diff
@@ -0,0 +1,272 @@
+import FWCore.ParameterSet.Config as cms
```
Comment (slava77) [inline Configuration/StandardSequences/python/ctppsRecoForSimulation2016_cff.py]: this looks like a temporary test config. it doesn't belong in  Configuration/StandardSequences main python dir.
please remove.

File: Configuration/StandardSequences/python/ctppsRecoForSimulation2017_cff.py [hunk 3]
```diff
@@ -0,0 +1,272 @@
+import FWCore.ParameterSet.Config as cms
```
Comment (slava77) [inline Configuration/StandardSequences/python/ctppsRecoForSimulation2017_cff.py]: this looks like a temporary test config. it doesn't belong in Configuration/StandardSequences main python dir.
please remove.

File: Configuration/StandardSequences/python/ctppsRecoForSimulation2018_cff.py [hunk 4]
```diff
@@ -0,0 +1,274 @@
+import FWCore.ParameterSet.Config as cms
```
Comment (slava77) [inline Configuration/StandardSequences/python/ctppsRecoForSimulation2018_cff.py]: this looks like a temporary test config. it doesn't belong in Configuration/StandardSequences main python dir.
please remove.

File: EventFilter/CTPPSRawToDigi/plugins/CTPPSPixelDigiToRaw.h [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/Utilities/interface/InputTag.h"
+#include "CondFormats/DataRecord/interface/CTPPSPixelDAQMappingRcd.h"
+#include "CondFormats/CTPPSReadoutObjects/interface/CTPPSPixelDAQMapping.h"
+#include "DataFormats/Common/interface/DetSetVector.h"
+#include "DataFormats/CTPPSDigi/interface/CTPPSPixelDigi.h"
+#include "CondFormats/CTPPSReadoutObjects/interface/CTPPSPixelFramePosition.h"
+
+
+class CTPPSPixelDigiToRaw final : public edm::EDProducer {
```
Comment (slava77) [inline EventFilter/CTPPSRawToDigi/plugins/CTPPSPixelDigiToRaw.h]: legacy module base is not allowed for new modules.
Please use one of the MT-aware modules
https://twiki.cern.ch/twiki/bin/view/CMSPublic/FWMultithreadedFrameworkModuleTypes

File: EventFilter/CTPPSRawToDigi/plugins/CTPPSPixelDigiToRaw.h [hunk 6]
```diff
@@ -0,0 +1,55 @@
+#ifndef CTPPSPixelDigiToRaw_H
```
Comment (slava77) [inline EventFilter/CTPPSRawToDigi/plugins/CTPPSPixelDigiToRaw.h]: .h file is not really needed for producers and the content can be merged to the corresponding .cc file

File: EventFilter/CTPPSRawToDigi/plugins/CTPPSTotemDigiToRaw.h [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+#include "CondFormats/DataRecord/interface/TotemReadoutRcd.h"
+#include "CondFormats/CTPPSReadoutObjects/interface/TotemDAQMapping.h"
+#include "DataFormats/Common/interface/DetSetVector.h"
+#include "DataFormats/CTPPSDigi/interface/TotemRPDigi.h"
+#include "DataFormats/CTPPSDetId/interface/TotemRPDetId.h"
+#include "CondFormats/CTPPSReadoutObjects/interface/TotemFramePosition.h"
+#include "EventFilter/CTPPSRawToDigi/interface/VFATFrameCollection.h"
+#include "DataFormats/CTPPSDigi/interface/TotemVFATStatus.h"
+
+class CTPPSTotemDigiToRaw final : public edm::EDProducer {
```
Comment (slava77) [inline EventFilter/CTPPSRawToDigi/plugins/CTPPSTotemDigiToRaw.h]: legacy module base is not allowed for new modules.
Please use one of the MT-aware modules
https://twiki.cern.ch/twiki/bin/view/CMSPublic/FWMultithreadedFrameworkModuleTypes

File: EventFilter/CTPPSRawToDigi/plugins/CTPPSTotemDigiToRaw.h [hunk 8]
```diff
@@ -0,0 +1,53 @@
+#ifndef CTPPSTotemDigiToRaw_H
```
Comment (slava77) [inline EventFilter/CTPPSRawToDigi/plugins/CTPPSTotemDigiToRaw.h]: .h file is not really needed for producers and the content can be merged to the corresponding .cc file

Please use an include guard matching the full file name subsystem_package_fileName_h

File: EventFilter/CTPPSRawToDigi/plugins/SealModule.cc [hunk 9]
```diff
@@ -0,0 +1,9 @@
+#include "FWCore/PluginManager/interface/ModuleDef.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+
+#include "EventFilter/CTPPSRawToDigi/plugins/CTPPSPixelDigiToRaw.h"
+#include "EventFilter/CTPPSRawToDigi/plugins/CTPPSTotemDigiToRaw.h"
+
+DEFINE_FWK_MODULE(CTPPSPixelDigiToRaw);
+DEFINE_FWK_MODULE(CTPPSTotemDigiToRaw);
```
Comment (slava77) [inline EventFilter/CTPPSRawToDigi/plugins/SealModule.cc]: SealModule.cc is not really needed.
The ```DEFINE_FWK_MODULE``` can be in the corresponding .cc file where the producers are defined

File: EventFilter/CTPPSRawToDigi/plugins/CTPPSPixelDigiToRaw.h [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+  void endJob() override {}
+
+  //static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+  /// get data, convert to raw event, attach again to Event
+  void produce( edm::Event&, const edm::EventSetup& ) override;
+typedef uint64_t Word64;
+
+private:
+  edm::ParameterSet config_;
```
Comment (slava77) [inline EventFilter/CTPPSRawToDigi/plugins/CTPPSPixelDigiToRaw.h]: please do not make copies of the full PSet. Simply extract all needed parameters from it in the module constructor.

File: EventFilter/CTPPSRawToDigi/plugins/CTPPSPixelDigiToRaw.h [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+
+
+  /// dummy end of job 
+  void endJob() override {}
+
+  //static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+  /// get data, convert to raw event, attach again to Event
+  void produce( edm::Event&, const edm::EventSetup& ) override;
+typedef uint64_t Word64;
```
Comment (slava77) [inline EventFilter/CTPPSRawToDigi/plugins/CTPPSPixelDigiToRaw.h]: please define a sensible fillDescriptions method 
https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideConfigurationValidationAndHelp

File: EventFilter/CTPPSRawToDigi/plugins/CTPPSTotemDigiToRaw.h [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+
+  /// dummy end of job 
+  void endJob() override {}
+
+  /// get data, convert to raw event, attach again to Event
+  void produce( edm::Event&, const edm::EventSetup& ) override;
+  typedef uint64_t Word64;
+
+private:
+  edm::ParameterSet config_;
```
Comment (slava77) [inline EventFilter/CTPPSRawToDigi/plugins/CTPPSTotemDigiToRaw.h]: please do not make copies of the full PSet. Simply extract all needed parameters from it in the module constructor.

File: EventFilter/CTPPSRawToDigi/plugins/CTPPSTotemDigiToRaw.h [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+  /// dtor
+  ~CTPPSTotemDigiToRaw() override;
+
+
+  /// dummy end of job 
+  void endJob() override {}
+
+  /// get data, convert to raw event, attach again to Event
+  void produce( edm::Event&, const edm::EventSetup& ) override;
+  typedef uint64_t Word64;
```
Comment (slava77) [inline EventFilter/CTPPSRawToDigi/plugins/CTPPSTotemDigiToRaw.h]: please define a fillDescriptions here as well

File: EventFilter/CTPPSRawToDigi/python/ctppsDigiToRaw_cfi.py [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+ctppsPixelRawData = cms.EDProducer("CTPPSPixelDigiToRaw",
+    InputLabel = cms.InputTag("RPixDetDigitizer"),
+    mappingLabel = cms.string("RPix") 
+)
+
+ctppsTotemRawData = cms.EDProducer("CTPPSTotemDigiToRaw",
+    InputLabel = cms.InputTag("RPSiDetDigitizer")
+)
+
+ctppsRawData = cms.Sequence(ctppsTotemRawData*ctppsPixelRawData)
```
Comment (slava77) [inline EventFilter/CTPPSRawToDigi/python/ctppsDigiToRaw_cfi.py]: sequences should not be defined in cfi files. Please move to cff.
Normally, a cfi file should define only one module instance with its default parameters.

For this file this means:
- ctppsPixelRawData_cfi and ctppsTotemRawData_cfi are needed . Note, that if the fillDescriptions methods are defined appropriately, these files will be generated by the framework
- this file can be renamed to ctppsRawData_cff or ctppsDigiToRaw_cff

Comment (slava77) [inline EventFilter/CTPPSRawToDigi/python/ctppsDigiToRaw_cfi.py]: it seems like the default should be an empty sequence and a non-empty should be defind with ctpps_2016

File: EventFilter/CTPPSRawToDigi/python/pgctppsDigiToRaw_cfi.py [hunk 15]
```diff
@@ -0,0 +1,10 @@
+import FWCore.ParameterSet.Config as cms
```
Comment (slava77) [inline EventFilter/CTPPSRawToDigi/python/pgctppsDigiToRaw_cfi.py]: move to cff file

File: EventFilter/CTPPSRawToDigi/python/pgctppsRawToDigi_cfi.py [hunk 16]
```diff
@@ -0,0 +1,10 @@
+import FWCore.ParameterSet.Config as cms
```
Comment (slava77) [inline EventFilter/CTPPSRawToDigi/python/pgctppsRawToDigi_cfi.py]: move to cff file

**Review by slava77 — COMMENTED** [review]
this is rather cursory, covering reco-related files.
Some more comprehensive notes can be expected some time next week (after the code runs)

**Comment by civanch** [comment]
@mundim , would it be possible to split this PR on some orthogonal peaces? Here you add full developments and it is difficult for reviewers even to access a particular file.

**Comment by mundim** [comment]
> would it be possible to split this PR on some orthogonal peaces? Here you add full developments and it is difficult for reviewers even to access a particular file.

Yes, of course. How you suggest to do it? I would point out 4 PRs that can for sure be split:
1 - PPSTools (new package)
2 - SimTransport (drastically changed package, following up the changes stated in Jan, 2017 
3 - FastSimulation/CTPPS* (no new feature, just adaptation of the old fast sim to the new transport
4 - The PPS fullsim (new and bigger chunk of code)
In the simulation meeting, this was the suggestion, but we were told we cold do it as we wish, since all the code was updated in the same branch and given the short time, we decided to do it at once. We can do the way it is better and easier to get it though.

**Comment by civanch** [comment]
@mundim , now when we far beyond 10_2 schedule there is no problem go step by step. You may try to start from 1 or 1+2 and see how it will pass review. I suspect, that this will be much faster.

**Comment by kpedro88** [comment]
@mundim you should make a branch with just the changes in that package, and then make a new PR. (Please make sure the new branch compiles before submitting the PR.)

File: Geometry/CMSCommonData/data/extend/cmsextent.xml [hunk 17]
```diff
@@ -13,15 +13,20 @@
 	</ConstantsSection>
 	<SolidSection label="cmsextent.xml">
 		<Tubs name="CMStoZDC" rMin="[Rmin]" rMax="[CMSR2]" dz="56*m" startPhi="0*deg" deltaPhi="360*deg"/>
-		<Tubs name="ZDCtoFP420" rMin="[Rmin]" rMax="[CMSR2]" dz="120*m" startPhi="0*deg" deltaPhi="360*deg"/>
+		<Tubs name="ZDCtoVF200" rMin="[Rmin]" rMax="[CMSR2]" dz="26*m" startPhi="0*deg" deltaPhi="360*deg"/>
```
Comment (ianna) [inline Geometry/CMSCommonData/data/extend/cmsextent.xml]: @mundim - you must create a new version of this file in order not to change previous geometries. The rule for new files is as follows _name/year/version/name.xml_:
```
Geometry/CMSCommonData/data/cmsextent/2018/v1/cmsextent.xml
```

**Review by ianna — CHANGES_REQUESTED** [review]

**Comment by ianna** [comment]
-1

**Comment by mundim** [comment]
[https://github.com/cms-sw/cmssw/pull/24420#issuecomment-418027754](url)
Since 1 and 2 are (almost) independent packages, I would suggest to make these first, then we can proceed with 3+4 together. Is this reasonable?

**Comment by mundim** [comment]
@civanch @kpedro88 Ok, I'll do 1+2 together...

**Comment by civanch** [comment]
@mundim , this is OK. I would advice to keep initial area from where this PR was pushed and create a fresh one based on the recent IB. You may copy files between areas, after 1+2 will be merged you may start 3 and 4 in other areas.

**Comment by mundim** [comment]
@civanch @kpedro88 

I prepared a new branch with the item 1+2+3 (the third step is just adaptation to the new transport and was already in the release, in case it is not ok, I can redo it). However, after git checkdeps -a, the compilation fails at 

RecoEcal/EgammaClusterProducers/src/EcalDigiSelector.cc 

because of:

template <class T> struct is_complex<std::complex<T> > : public true_type{}: complex is not a member of std

The packages for which I'm issuing the PR doesn't depend on this package, so I don't know what to do.

**Comment by mundim** [comment]
Just a clarification: this PR needs to be closed and reissued again due to the PR splitting (PR 24490) or will it proceed normally?

**Comment by kpedro88** [comment]
This PR should be closed in favor of the split PRs.

**Comment by mundim** [comment]
Then, a new PR should be issued. We understood it wrong. Is there time yet to issue the PR for the remaining code? Sorry about it.

**Comment by kpedro88** [comment]
There is still a week until the last 10_3_X prerelease is closed (the schedule was revised recently), so you can start preparing the subsequent PR.

**Comment by mundim** [comment]
Closing this PR, as agreed when asked to split this PR in smaller ones.

**Pull request closed without merging**